core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c

```diff
@@ -1085,6 +1118,181 @@ posix_memalign_wrapper(wasm_exec_env_t exec_env, void **memptr, int32 align,
     return 0;
 }
 
+#if WASM_ENABLE_LIB_PTHREAD_SEMAPHORE != 0
+
+static int32
+sem_open_wrapper(wasm_exec_env_t exec_env, const char *name, int32 oflags,
+                 int32 mode, int32 val)
+{
+    korp_sem *psem = NULL;
+    ThreadInfoNode *info_node = NULL;
+
+    /**
+     * For RTOS, global semaphore map is safe for share the same semaphore
+     * between task/pthread.
+     * For Unix like system, it's dedicated for multiple processes.
+     */
+
+    if ((info_node = bh_hash_map_find(sem_info_map, (void *)name))) {
+        return info_node->handle;
+    }
+
+    if (!(psem = os_sem_open(name, oflags, mode, val))) {
+        goto fail1;
+    }
+
+    if (!(info_node = wasm_runtime_malloc(sizeof(ThreadInfoNode))))
+        goto fail2;
+
+    memset(info_node, 0, sizeof(ThreadInfoNode));
+    info_node->exec_env = exec_env;
+    info_node->handle = allocate_handle();
+    info_node->type = T_SEM;
+    info_node->u.sem = psem;
+    info_node->status = SEM_CREATED;
+
+    if (!append_thread_info_node(info_node)) {
+        wasm_runtime_free(info_node);
+        goto fail2;
+    }
+
+    if (!bh_hash_map_insert(sem_info_map, (void *)name, info_node))
+        goto fail3;
+
+    return info_node->handle;
+
+fail3:
+    delete_thread_info_node(info_node);
+fail2:
+    os_sem_close(psem);
+fail1:
+    return -1;
+}
+
+void
+sem_fetch_cb(void *key, void *value, void *user_data)
+{
+    (void)key;
+    SemCallbackArgs *args = user_data;
+    ThreadInfoNode *info_node = value;
+    if (args->handle == info_node->handle && info_node->status == SEM_CREATED) {
+        args->node = info_node;
+    }
+}
+
+static int32
+sem_close_wrapper(wasm_exec_env_t exec_env, uint32 sem)
+{
+    (void)exec_env;
+    int ret = -1;
+    SemCallbackArgs args = { sem, NULL };
+
+    bh_hash_map_traverse(sem_info_map, sem_fetch_cb, &args);
```

Comment:
Traversing the sem_info_map is slow, as it traverses the hash map elements one by one. Had better call get_thread_info like other place, as it uses handle as the key to retrieve hash table, the performance is better. Seems the code can be like:
```C
ThreadInfoNode *node;
...
node = get_thread_info(exec_env, sem);
if (node && node->type == T_SEM && node->status == SEM_CREATED) {
    ret = os_sem_close(node->u.sem);
    ...
}
```

Reply:
Yes, it's slow but by this comment I found that we should not append this node into thread_info_map since it will be destored twice , so we can get the info node by get_thread_info.

Reply:
OK, it depends you. In our original design, all the resources created by a thread should be destroyed when the thread exits.
The resource might be destroyed twice, since you set `thread_info_destroy` for the value destroying callback when creating sem_info_map:
```C
sem_info_map = bh_hash_map_create(
              32, true, (HashFunc)wasm_string_hash,
              (KeyEqualFunc)wasm_string_equal, NULL, thread_info_destroy)
```
thread_info_destroy was already set for creating thread_info_map:
```C
node->thread_info_map = bh_hash_map_create(
              32, true, (HashFunc)thread_handle_hash,
              (KeyEqualFunc)thread_handle_equal, NULL, thread_info_destroy)
```
So it may be called in both destroying thread_info_map and destroying sem_info_map. 
There might be no double free in the sample, as it unlink the sem, and the sem was removed from sem_hash_map in sem_unlink_wrapper.

So here we can just set NULL when creating sem_info_map, if we want to append sem into thread_info_map and avoid double free:
```C
sem_info_map = bh_hash_map_create(
              32, true, (HashFunc)wasm_string_hash,
              (KeyEqualFunc)wasm_string_equal, NULL, NULL);
```